File: Programmers/N으로 표현.py
def solution(N, number):
    answer = -1
    DP = []

    for i in range(1, 9):
        count_set = set()
        count_set.add(int(str(N) * i))
        for j in range(i - 1):
            for x in DP[j]:
                for y in DP[-j - 1]:
                    count_set.add(x + y)
                    count_set.add(x - y)
                    count_set.add(x * y)

                    if y != 0:
                        count_set.add(x // y)
        if number in count_set:
            answer = i
            break;
        DP.append(count_set)

    return answer

File: Programmers JS/N으로 표현.js
function solution(N, number) {
  var answer = 0;
  const dp = Array.from({ length: 9 }, () => new Set());
  if (N === number) return 1;

  dp.forEach((set, i) => {
    if (i !== 0) set.add(Number(N.toString().repeat(i)));
  });

  for (let i = 1; i < 9; i++) {
    for (let j = 1; j < i; j++) {
      for (let post of dp[j]) {
        for (let prev of dp[i - j]) {
          dp[i].add(post + prev);
          dp[i].add(post - prev);
          dp[i].add(post * prev);
          dp[i].add(Math.floor(post / prev));
        }
      }
      if (dp[i].has(number)) return i;
    }
  }
  return -1;
}
